doc/comptime.org

```diff
@@ -0,0 +1,495 @@
+
+  #define DEBUG 1
+
+  #if DEBUG 
+  #define TRACE(x) printf x;
+  #else
+  #define TRACE(x)
+  #endif
+
+  int main() {
+    TRACE(("Hello World! : %d\n", 12));
+  }
+#+end_src
+
+#+RESULTS:
+: Hello World! : 12
+
+In Zig, logging uses this same principle, a message level is set at the start of the program (at compile-time) and if the log is not enabled, all the code about the print is removed.
+However, if the log level is greater than the limit, the message will be printed.
+
+The code below shows an extract of the standard library for logging.
+#+begin_src zig
+ fn log(
+    comptime message_level: Level,
+    comptime scope: @Type(.EnumLiteral),
+    comptime format: []const u8,
+    args: anytype,
+) void {
+    if (comptime !logEnabled(message_level, scope)) return;
+
+    std.options.logFn(message_level, scope, format, args);
+}
+#+end_src
+
+In addition, Zig provides some helper functions for logging, such as :
+- ~std.log.debug~
+- ~std.log.info~
+- ~std.log.warn~
+- ~std.log.err~
+
+And if the release mode is set to ~Debug~, the debug log will be printed.
+But if the release mode is set to ~Release*~, the debug log will not print, there is no need to configure the logging to have this behavior.
+
+** Generic data structures
+To create a generic data structure, the same pattern is used as a comptime parameter.
+A function needs to return an anonymous struct as a type ~type~.
+
+In a generic data structure, the ~@This()~ is used to get the type of the data structure because she is anonymous.
+
+Moreover, with generic data structure can have a function that can be called on the structure type and can have a function call on the instance of the structure.
+To have an instance function, the first argument needs to be a parameter of the type of the struct.
+That's why a constant ~Self~ is used with ~@This()~.
+And after that, the parameter ~self~ can be used to get the members of the struct.
+
+The example shows the difference between a function that can be called on a struct and a function that can be called on an instance of a struct.
+#+begin_src zig :imports '(std) :main 'no
+  pub fn MyStruct(comptime T: type) type {
+      return struct {
+          const Self = @This();
+
+          myNumber: T,
+
+          pub fn structFunction(writer: anytype) !void {
+              try writer.print("structFunction\n", .{});
+          }
+
+          pub fn instanceFunction(self: *Self, writer: anytype) !void {
+              try writer.print("structInstance: {d}\n", .{self.myNumber});
+          }
+      };
+  }
+
+  pub fn main() !void {
+      const stdout = std.io.getStdOut().writer();
+
+      try MyStruct(f32).structFunction(stdout);
+
+      var myStruct = MyStruct(f32){
+          .myNumber = 42,
+      };
+
+      try myStruct.instanceFunction(stdout);
+  }
+
+#+end_src
+
+#+RESULTS:
+| structFunction  |    |
+| structInstance: | 42 |
+
+Zig can infer the name of a struct when there are created, or a name can be explicitly given :
```

Comment:
Not sure to understand what you want to put forward by saying that.

Reply:
It's more clear ?